### DISCUSSION TITLE
Help wanted - NPC behaviour trees

### DISCUSSION THREAD

**Discussion by ErinaSugino** [ErinaSugino/Starbound-SBR-Mod#41] [General]
Looking for people with the necessary knowledge to understand the inner workings of NPC behaviour trees. That stuff seems like black magic to me, but I'll be needing to work with it for a variety of planned features.

If you have experience in working with them, or even better know of a tool that I can use to work with them more easily, please let me know.

**Comment by XspeedPL** [comment]
I can help, but please try to be more specific

**Reply by ErinaSugino** [reply]
I'm afraid I cannot be much more specific than that, as I already lack the basic understanding of it's core concept.

I know it's a "tree" logic - "nodes" are executed after eachother, sometime splitting into sub-routines, which do *something somehow* and then even share data with eachother. What's the syntax? Which modules exist, how do they work?

To take a more reality grounded usecase, let's talk about SBR. In order for NPCs to have an "arousal" mechanic, they have (alongside all the other SxB-related behaviours) a behaviour script patched into their core loop. This checks if their "arousal" resource is 100, then executes a loop which fetches all objects classified as "sexnode" in range, filters them out when they are "full", sorts them by distance and then walks over to the nearest one. These checks are done by calling "action" methods to be found in a seperate script of Sexbound, which contains wrapper methods which then call the actual instance (if any) on the object, receiving some arbitrary environment variables as parameters (game board something something), then returns data.

How does all of this work? How would I go about expanding this logic to add more sophisticated state checks to this logic, to make the behaviour of choosing a node more logical? To be specific in this single case, I want to add information that prevents incest (data on entities, checked by the Sexnode) and make NPCs prioritize nodes that already contain another actor of another gender, if possible. That would require forwarding data of the NPC currently "behaving", sending it to the sexnode and then receiving a response, which is used in further "decisions".
